doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,796 @@
+would then find correct dtype instance (the second pass is often not necessary).
+The advantage of this is that the DType class information is vital for universal
+functions to decide which loop to execute.
+The first pass would provide the full information necessary for value based
+casting currently implemented for scalars, giving even the possibility of
+expanding it to e.g. list inputs ``np.add(np.array([8], dtype="uint8"), [4])``
+giving a ``uint8`` result.
+This is mainly related to the question to how the common dtype is found above.
+
+
+Common DType Operations
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Numpy currently provides functionality of ``np.result_type`` and
+``np.promote_types`` (while more common, ``np.result_type`` has some more
+complex logic due to implementing value based promotion [value_based]_).
+(Note that the name ``np.common_type`` is associated with the scalars and not
+directly with dtypes and is of limited usefulness)
+
+To distinguish between the promotion occuring during universal function application,
+here common type operation is used instead of promotion.
+
+**Motivation:**
+This common type operations is vital for the above array coercion when different
+input types are mixed.
+It also provides the logic currently used to decide the output dtype of
+``np.concatenate()`` and on its own is a useful operation.
+
+Furthermore, it may be used to find the correct dtype to use for functions with
+different inputs (including universal functions).
+This includes an interesting distinction:
+
+1. Universal functions use the DType classes for dispatching, they thus
+   require the common DType class (as a first step).
+   While this can help with finding the correct loop to execute, the loop
+   may not need the actual common dtype instance.
+   (Hypothetical example:
+   ``float_arr + string_arr -> string``, but the output string length is
+   not the same as ``np.concatenate(float_arr, string_arr)).dtype``.)
+2. Array coercion, and concatenate require the common dtype *instance*.   
+
+**Implementation:**
+The implementation of the common dtype (instance) has some overlap with
+casting.
+Casting from a specific dtype (Float64) to a String needs to find
+the correct string length (a step that is mainly necessary for flexible dtypes).
+
+We propose the following implementation:
+
+1. ``__common_dtype__(cls, other : DTypeMeta) -> DTypeMeta`` answers what the common
+   DType class is. It may return ``NotImplemented`` to defer to ``other``.
+   (For abstract DTypes, subclasses get precedence, concrete types are always
+   leaves, so always get preference or are tried from left to right). 
+2. ``__common_instance__(self, other : cls) -> cls`` is used when two instances
+   of the same DType are given. For builtin dtypes (that are not flexible), this
+   currently always returns ``self`` (but ensures native byte order).
+   This is to preserve metadata. We can thus provide a default implementation
+   for non-flexible user dtypes.
```

Comment:
`__len__` is only generally defined for a DType instance! Abstract dtypes cannot be instantiated. Of course a `Float128` could inform you that all instances are 128bit, but there is no reason it has to.

I am not sure about `dtype1 & dtype2`. It might make sense to note such possibilities at some point though, if only to discourage downstream to implement dtypes which use such API for something else.

Reply:
Did not change anything here.
About the `__call__` I think it would be OK for an abstract DType to actually return a concrete DType instance which it considers a subclass of itself (i.e. similar to a class factory). Creating the scalar should rather be through `DType.type()` or another classmethod I think.

The reuse of existing slots is interesting... We could just do that, or at least document that we may want to do that at some point ("please do not fill them in", we could even enforce that most operators are not defined by users, e.g. we have authority over what `==` does).

The last part should maybe be added, even if it is a detail?

Reply:
Interesting ideas - I think it would be good to add that we are still considering to what extent existing slots can be useful.